### ISSUE TITLE
Support structured logging

### ISSUE THREAD

**Issue by Alxandr** [eventflow/EventFlow#591]
The current logging in EventFlow is based on `String.Format` (or just plain strings). This means that structural logging frameworks like Serilog can't get any properties out of it (except maybe if you extract properties with names like 0, 1, 2, etc.). I've written an `ILog` adapter for `Microsoft.Extensions.Logging` (which I've then connected to Serilog), and as you can see, I don't get any extracted information from the eventflow log messages (it's just one big string):

![image](https://user-images.githubusercontent.com/112334/53292182-b2d38a80-37bf-11e9-81fa-1e6af386586c.png)

I would recommend switching to the `Microsoft.Extensions.Logging` as the log abstraction framework, as that's what you get out of the box with Asp.NET Core, so chances are most logging frameworks will support it.

**Comment by rasmus** [comment]
An important cleanup, the only problem is to minimize breaking changes. Even though EventFlow haven't shipped a 1.x version yet, I still work hard not to do any breaking changes.

**Comment by Alxandr** [comment]
Creating a bridge from the `Microsoft.Extensions.Logging` abstraction to the `ILog` one present today should be pretty straight forward. Log messages in the extentions logging framework is required to have a `formatter` delegate passed around that knows how to stringify the objects. It prefectly translates to the model `ILog` uses (with the loss of the structure of the data). Then it should just be a matter of checking if any custom `ILog` implementations has been registered, and if so wire up the bridge I would think.

**Comment by rasmus** [comment]
Yes, but pulling in `Microsoft.Extensions.Logging` instead of `Microsoft.Extensions.Logging.Abstractions` also pulls in `Microsoft.Extensions.Configuration`, `Microsoft.Extensions.Options` and `Microsoft.Extensions.DependencyInjection`. If these packages are referenced from EventFlow I (as an any user of EventFlow) would expect EventFlow to fully use these packages and respect their conventions.

- Allow configuration using options
- Use Microsoft DI as default

Alternatively EventFlow could merely reference `Microsoft.Extensions.Logging.Abstractions`, but that leaves doing the implementation of all the structure logging implementation details you get in `Microsoft.Extensions.Logging` to EventFlow itself as all internal logs should be rewritten to be structured logs, but still be backward compatible with the `ILog` interface which current implementations doesn't necessarily support structure log messages.

So while the structure log in themselves are pretty strait forward, simply by referencing the new NuGet packages does pose additional questions and complexity.

I do like the idea of using the `Microsoft.Extensions.*` packages as it would allow some code to be deleted from EventFlow, e.g. the custom IoC. But those packages are sort of a "package deal".

Another potential problem, is that the requirement for EventFlow on .NET Framework projects would be 4.7.something as opposed to 4.5.1, which might be a deal breaker for some developers.

**Comment by Alxandr** [comment]
I'm pretty sure you can make due with the abstractions package, and a fairly minimal bridge from the new ILogger interface to the old ILog interface. But yes, if it was up to me I would take in most of the Extensions packages (like DI, etc.). I've written a wrapper for EventFlow allowing me to use the MS DI and logging. But I figured that was way out of scope, so I kept the suggestion to "only" structured logging.

The `ILogger` interface is really quite simple with only 3 methods ([link](https://github.com/aspnet/Extensions/blob/master/src/Logging/Logging.Abstractions/src/ILogger.cs)). Just of the top of my head (with no editor help), the main method would be something like:

```csharp
public class LogWrapper : ILogger
{
  private ILog _log;

  public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter) {
    switch(logLevel) {
      case LogLevel.Information:
        _log.Information(formatter(state, exception));
        break;
      ......
    }
  }
}
```

**Comment by rasmus** [comment]
Have a look at 
https://github.com/aspnet/Logging/blob/master/src/Microsoft.Extensions.Logging.Abstractions/Internal/FormattedLogValues.cs

**Comment by Alxandr** [comment]
@rasmus https://www.nuget.org/packages/Microsoft.Extensions.Logging/2.2.0 supports netstandard2...

**Comment by rasmus** [comment]
This is implemented in 1.0
